include/volume.h

```diff
@@ -0,0 +1,276 @@
+        if (print) std::cout<<"---------"<<std::endl;
+
+        // 5. Estimate Vol(P)
+
+        NT telescopic_prod=NT(1);
+
+        std::vector<Ball>::iterator bit2=balls.end();
+        bit2--;
+
+        while(bit2!=balls.begin()){
+
+            //each step starts with some random points in PBLarge stored in list "randPoints"
+            //these points have been generated in a previous step
+
+            BallPoly PBLarge(P,*bit2);
+            --bit2;
+            BallPoly PBSmall(P,*bit2);
+
+            if(print)
+                std::cout<<"("<<balls.end()-bit2<<"/"<<balls.end()-balls.begin()<<") Ball ratio radius="
+                        <<PBLarge.second().radius()<<","<<PBSmall.second().radius()<<std::endl;
+
+            // choose a point in PBLarge to be used to generate more rand points
+            Point p_gen = *randPoints.begin();
+
+            // num of points in PBSmall and PBLarge
+            int nump_PBSmall = 0;
+            int nump_PBLarge = randPoints.size();
+
+            if(print) std::cout<<"Points in PBLarge="<<randPoints.size()
+                              <<std::endl;
+
+            //keep the points in randPoints that fall in PBSmall
+            std::list<Point>::iterator rpit=randPoints.begin();
+            while(rpit!=randPoints.end()){
+                if (PBSmall.second().is_in(*rpit) == 0){//not in
+                    rpit=randPoints.erase(rpit);
+                } else {
+                    ++nump_PBSmall;
+                    ++rpit;
+                }
+            }
+
+            if(print) std::cout<<"Points in PBSmall="<<randPoints.size()
+                              <<"\nRatio= "<<NT(nump_PBLarge)/NT(nump_PBSmall)
+                             <<std::endl;
+
+            if(print) std::cout<<"Generate "<<rnum-nump_PBLarge<<  " more "
+                              <<std::endl;
+
+            //generate more random points in PBLarge to have "rnum" in total
+            rand_point_generator(PBLarge,p_gen,rnum-nump_PBLarge,walk_len,randPoints,PBSmall,nump_PBSmall,var);
+
+            telescopic_prod *= NT(rnum)/NT(nump_PBSmall);
+            if(print) std::cout<<nump_PBSmall<<"/"<<rnum<<" = "<<NT(rnum)/nump_PBSmall
+                              <<"\ncurrent_vol="<<telescopic_prod
+                             <<"\n="<<telescopic_prod
+                            <<"\n--------------------------"<<std::endl;
+
+            //don't continue in pairs of balls that are almost inside P, i.e. ratio ~= 2
+        }
+        
+        if(print) std::cout<<"rand points = "<<rnum<<std::endl;
+        if(print) std::cout<<"walk len = "<<walk_len<<std::endl;
+        vol = (std::pow(M_PI,n/2.0)*(std::pow(balls[0].radius(), n) ) ) / (tgamma(n/2.0+1));
+        vol=vol*telescopic_prod;
+        if(print) std::cout<<"round_value: "<<round_value<<std::endl;
+        vol=round_value*vol;
+        if(print) std::cout<<"volume computed: "<<vol<<std::endl;
+
+    }
+
+    return vol;
+}
+
+#endif
```

Comment:
add newline at the end of file (do it in the whole project if applicable)

Reply:
done